cmd/unifibackup/main.go

```diff
@@ -32,73 +32,56 @@ var (
+	} `cmd:"" help:"Copies UniFi Controller backups to S3" default:"withargs"`
+}
+
 func main() {
 	maxprocs.Set() // use the library this way to avoid logging when CPU quota is undefined
 	buildInfo.WithLabelValues(stamp.Version, stamp.Commit).Set(1)
 	buildTime.Set(float64(stamp.Time().UnixNano()) / float64(time.Second))
 
-	// we can access flags via cmd, but they are untyped; in practice it's
-	// easier to constrain the chaos to this function
-	var (
-		flgBackupDir string
-		flgBucket    string
-		flgPrefix    string
-		flgMetrics   string
-		flgTimeout   time.Duration
-	)
-
-	rootCmd := &cobra.Command{
-		Use:          "unifibackup --bucket unifi-backups",
-		Short:        "Copies UniFi Controller backups to S3",
-		Version:      stamp.Summary(),
-		SilenceUsage: true,
-		Args:         cobra.NoArgs,
-		RunE: func(cmd *cobra.Command, _ []string) error {
-			ctx := cmd.Context()
-			cfg, err := config.LoadDefaultConfig(ctx,
-				config.WithUseDualStackEndpoint(aws.DualStackEndpointStateEnabled))
-			if err != nil {
-				return fmt.Errorf("failed to initialise AWS SDK: %w", err)
-			}
-			s3client := s3.NewFromConfig(cfg)
-			uploader := uploader.New(s3client, flgBucket, flgPrefix)
-			return daemon(ctx, flgMetrics, flgBackupDir, uploader, flgTimeout)
-		},
-	}
-	rootCmd.SetVersionTemplate("{{.Version}}\n")
-	rootCmd.PersistentFlags().StringVar(&flgBackupDir,
-		"dir",
-		"/var/lib/unifi/backup/autobackup",
-		"path of the UniFi autobackup directory")
-	rootCmd.Flags().StringVar(&flgBucket,
-		"bucket",
-		"", // in hindsight, this should not have been a flag
-		"name of the S3 bucket to upload to (requird)")
-	rootCmd.MarkFlagRequired("bucket")
-	rootCmd.Flags().StringVar(&flgPrefix,
-		"prefix",
-		"unifi/",
-		"prepended to the backup file name to form the object key")
-	rootCmd.Flags().StringVar(&flgMetrics,
-		"metrics",
-		":9184",
-		"listen spec for the web server that exposes Prometheus metrics")
-	rootCmd.Flags().DurationVar(&flgTimeout,
-		"timeout",
-		5*time.Minute,
-		"time to allow for upload and delete requests")
-
-	rootCmd.AddCommand(&cobra.Command{
-		Use:   "genmeta",
-		Short: "Generate autobackup_meta.json for the autobackup directory",
-		Args:  cobra.NoArgs,
-		RunE: func(_ *cobra.Command, _ []string) error {
-			return genmeta(flgBackupDir)
+	ctx := kong.Parse(
+		&CLI,
+		kong.Name("unifibackup"),
+		kong.DefaultEnvars("UNIFIBACKUP"),
+		kong.Description("Copies UniFi Controller backups to S3"),
+		kong.Vars{
+			"version": stamp.Summary(),
 		},
-	})
-
-	if err := rootCmd.Execute(); err != nil {
-		// error already printed (SilenceErrors == false)
-		os.Exit(1)
+	)
+	switch ctx.Command() {
+	case "backup":
+		ctx := context.Background()
+		cfg, err := config.LoadDefaultConfig(ctx,
+			config.WithUseDualStackEndpoint(aws.DualStackEndpointStateEnabled))
+		if err != nil {
+			panic(fmt.Errorf("failed to initialise AWS SDK: %w", err))
```

Comment:
It's a good idea not to `panic()`, perhaps it would be worth copying the pattern [here][] and returning the error from this function instead?

[here]: https://github.com/gebn/tflcycles_exporter/blob/main/cmd/tflcycles_exporter/main.go#L37-L44

Reply:
i mean printing the error and exiting with code 1 is the same as panic
but sure, i can switch it to the same pattern

Reply:
Ack, it's more about having `main()` be the (single) point of exit - knowing all other functions will return in all cases simplifies the mental model, and means we can choose how to handle those errors at the caller rather than callee.

Reply:
yep yep, very good practice. I didn't see a setup like that with cobra so was trying to stick with what was already done.